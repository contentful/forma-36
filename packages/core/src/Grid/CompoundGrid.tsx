import { Grid as OriginalGrid } from './Grid';
import { GridItem } from './GridItem/GridItem';

type CompoundGrid = typeof OriginalGrid & {
  Item: typeof GridItem;
};

export const Grid = OriginalGrid as CompoundGrid;
Grid.Item = GridItem;
