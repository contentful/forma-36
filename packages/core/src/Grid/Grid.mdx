[CSS Grid](https://developer.mozilla.org/en-US/docs/Glossary/Grid) based React component, comes with predefined values to ensure design consistency, and ease of use.

## Table of contents

- [How to use Grid](#how-to-use-grid)
- [Code examples](#code-examples)

## How to use Grid

The Grid consists of two components:

- Grid: Used as a container for GridItems
- GridItem: Elements within the Grid

<br />

## Code examples

### columns

When defined as a number, it will split the space into equally sized columns; it also accepts any of the [grid-template-columns](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-columns) css properties. e.g.

```jsx
<React.Fragment>
  <Grid columns={6}></Grid>
  <Grid columns={'auto 1fr'}></Grid>
</React.Fragment>
```

### rows

Accepts a number or any of [grid-template-rows](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-rows) css properties.

```jsx
<React.Fragment>
  <Grid rows={6}></Grid>
  <Grid rows={'auto 1fr'}></Grid>
</React.Fragment>
```

### columnGap & rowGap

`columnGap` represents space between columns, `rowGap` represents the space between rows, both accepts one of [spacing](https://f36.contentful.com/foundation/spacing/) token names.

`spacing2Xs` | `spacingXs` | `spacingXs` | `spacingS` | `spacingM` | `spacingL` | `spacingXl` | `spacing2Xl` | `spacing3Xl` | `spacing4Xl`

**e.g.**

```jsx
<Grid columnGap="spacingXs" rowGap="spacingXl"></Grid>
```
