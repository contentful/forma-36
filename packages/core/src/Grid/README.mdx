---
title: 'Grid'
type: 'component'
section: 'layoutComponents'
status: 'stable'
slug: /components/grid/
github: 'https://github.com/contentful/forma-36/tree/master/packages/core/src/Grid'
typescript: ./Grid.tsx,./GridItem/GridItem.tsx
---

`Grid` is a [CSS Grid](https://developer.mozilla.org/en-US/docs/Glossary/Grid) based React component, comes with predefined values to ensure design consistency, and ease of use.

The `Grid` consists of two components:

- `Grid`: Used as a container for grid items
- `Grid.Item`: Elements within the Grid

## Import

```js static=true
import { Grid } from '@contentful/f36-components';
```

## Grid vs. Flexbox

The main difference between CSS Grid Layout and CSS Flexbox Layout is that flexbox was designed for layouts in one dimension - either a row or a column.
Grid was designed for two-dimensional layout rows, and columns.

![One dimensional columns without rows describing the use flexbox, and two dimensional grid describes how the grid works](/docs-images/grid_vs_flexbox.png)

## Examples

### Basic usage

```jsx file=examples/BasicGridExample.tsx

```

### Repeat notation

`repeat()` is a css notation that you can use with the `gridColumns` and `girdRows` properties to make your rules more concise and easier to understand when creating a large amount of columns or rows

**Example:** `gridColumns="1fr 2fr 1fr 2fr 1fr 2fr"` is same as `gridColumns="repeat(3, 1fr 2fr)"` see below.

```jsx file=examples/RepeatNotationGridExample.tsx

```

### FR unit

Fr is a fractional unit and `1fr` stands for 1 part of the available space. It differs from percentage unit because it distributes available space.

If you place a larger item into a track then the way the fr until will work is to allow that track to take up more space and distribute what is left over.

```jsx file=examples/FrGridExample.tsx

```

### Span keyword

Span keyword works on grid items with the `gridArea` property, use `span` to avoid specifying the end lines for a column or a row.

i.e. instead of specifying `columnStart` and `columnEnd` you can use span to make the grid item take 4 columns

```jsx file=examples/SpanKeywordGridExample.tsx

```

### Content alignment

Just like [Flex] the `Grid` comes with `justifyContent`, `justifyItems`, `alignItems`, `alignContent`, ..etc.

```jsx file=examples/ContentAlignmentGridExample.tsx

```

## Accessibility

According to [WCAG C27](https://www.w3.org/TR/WCAG20-TECHS/C27.html) technique, the dom order must match the visual order.

The grid can re-order it's content in various ways through the order css property on grid items, this changes the visual order but the dom order stays the same, hence,
keyboard navigation and focus order maybe affected.

Set the correct order of the elements with `tabindex` attribute to avoid accessibility issues.

## Props (API reference)

### Grid

<Props of="Grid" storybookPath="/story/layout-grid--basic" />

### Grid.Item

<Props of="GridItem" />

[box]: ../components/box
[flex]: ../components/flex
