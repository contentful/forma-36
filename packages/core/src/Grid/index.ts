export { Grid } from './CompoundGrid';
export type { GridProps } from './Grid';
export { GridItem } from './GridItem/GridItem';
export type { GridItemProps } from './GridItem/GridItem';
