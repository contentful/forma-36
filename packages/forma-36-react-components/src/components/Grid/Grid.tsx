import React from 'react';
import tokens from '@contentful/forma-36-tokens';
import cn from 'classnames';
import type * as CSS from 'csstype';

import styles from './Grid.css';

export type GapTypes =
  | 'none'
  | 'spacing2Xs'
  | 'spacingXs'
  | 'spacingS'
  | 'spacingM'
  | 'spacingL'
  | 'spacingXl'
  | 'spacing2Xl'
  | 'spacing3Xl'
  | 'spacing4Xl';

export interface GridProps {
  /**
   * Class names to be appended to the className prop of the component */
  className?: string;
  /**
   * Child nodes to be rendered in the component */
  children?: React.ReactNode;
  /**
   * An ID used for testing purposes applied as a data attribute (data-test-id) */
  testId?: string;
  /**
   * Defines how many columns, default is `auto` */
  columns?: number | CSS.Property.GridTemplateColumns;
  /**
   * Defines how many rows, default is `auto` */
  rows?: number | CSS.Property.GridTemplateColumns;
  /**
   * Spaces between rows, corresponds to of spacing tokens values, default is none */
  rowGap?: GapTypes;
  /**
   * One of Spacing tokens values, default is 0 */
  columnGap?: GapTypes;
  /**
   * One of grid-auto-flow css values */
  flow?: CSS.Property.GridAutoFlow;
  /**
   * Sets display:inline-grid */
  inline?: boolean;
  /**
   * One of justify-content css values */
  justifyContent?: CSS.Property.JustifyContent;
  /**
   * One of justify-content css values */
  alignContent?: CSS.Property.AlignContent;
  /**
   * style prop, for inline styles */
  style?: React.CSSProperties;
}

const defaultProps = {
  columns: 'auto',
  rows: 'auto',
  columnGap: 'spacingM',
  rowGap: 'none',
  testId: 'cf-ui-grid',
};

export const Grid = (props: GridProps) => {
  const {
    className,
    children,
    testId,
    rows,
    columns,
    inline,
    rowGap,
    columnGap,
    justifyContent,
    alignContent,
    flow,
    style,
    ...otherProps
  } = props;

  const handleGridTemplate = (value?: string | number) => {
    if (typeof value === 'number') {
      return `repeat(${value}, minmax(0, 1fr)`;
    }
    return value;
  };

  const handleGap = (value: GapTypes) => {
    if (value === 'none') {
      return 0;
    } else {
      return tokens[value];
    }
  };

  const inlineStyle = {
    gridTemplateColumns: handleGridTemplate(columns),
    gridTemplateRows: handleGridTemplate(rows),
    flow,
    justifyContent,
    alignContent,
    columnGap: columnGap && handleGap(columnGap),
    rowGap: rowGap && handleGap(rowGap),
    ...style,
  };

  const classNames = cn(styles.Grid, className, {
    [styles.Grid__inline]: inline,
  });

  return (
    <div
      {...otherProps}
      style={inlineStyle}
      className={classNames}
      data-test-id={testId}
    >
      {children}
    </div>
  );
};
Grid.defaultProps = defaultProps;

export default Grid;
