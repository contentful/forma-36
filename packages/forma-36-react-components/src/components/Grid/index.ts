export { Grid } from './Grid';
export type { GridProps } from './Grid';
export { GridItem } from './GridItem/GridItem';
export type { GridItemProps } from './GridItem/GridItem';
