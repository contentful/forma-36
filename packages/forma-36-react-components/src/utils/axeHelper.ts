import { configureAxe } from 'jest-axe';

const axe = configureAxe({
  rules: {
    region: { enabled: false },
  },
});

export default axe;
