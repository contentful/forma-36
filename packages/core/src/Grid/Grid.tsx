import * as React from 'react';
import { css, cx } from 'emotion';
import {
  PolymorphicProps,
  PolymorphicComponent,
  ExpandProps,
} from '../Primitive/Primitive';
import { useBox } from '../Box';
import type * as CSS from 'csstype';
import type { MarginProps, PaddingProps, CommonProps, Spacing } from '../types';
import { convertSpacingToToken } from '../utils/getSpacingStyles';

export interface GridInternalProps
  extends CommonProps,
    MarginProps,
    PaddingProps {
  /**
   * Child nodes to be rendered in the component */
  children?: React.ReactNode;
  /**
   * Defines how many columns, default is `auto` */
  columns?: number | CSS.Property.GridTemplateColumns;
  /**
   * Defines how many rows, default is `auto` */
  rows?: number | CSS.Property.GridTemplateColumns;
  /**
   * Spaces between rows, corresponds to of spacing tokens values, default is none */
  rowGap?: Spacing;
  /**
   * One of Spacing tokens values, default is 0 */
  columnGap?: Spacing;
  /**
   * One of grid-auto-flow css values */
  flow?: CSS.Property.GridAutoFlow;
  /**
   * Sets display:inline-grid */
  isInline?: boolean;
  /**
   * One of justify-content css values */
  justifyContent?: CSS.Property.JustifyContent;
  /**
   * One of justify-content css values */
  alignContent?: CSS.Property.AlignContent;
}

const GRID_DEFAULT_TAG = 'div';

export type GridProps<
  E extends React.ElementType = typeof GRID_DEFAULT_TAG
> = PolymorphicProps<GridInternalProps, E>;

function _Grid<E extends React.ElementType = typeof GRID_DEFAULT_TAG>(
  {
    alignContent,
    children,
    columnGap = 'spacingM',
    columns = 'auto',
    flow,
    isInline,
    justifyContent,
    rowGap = 'none',
    rows = 'auto',
    as,
    ...otherProps
  }: GridProps<E>,
  ref: React.Ref<any>,
) {
  const handleGridTemplate = (value?: string | number) => {
    if (typeof value === 'number') {
      return `repeat(${value}, minmax(0, 1fr))`;
    }
    return value;
  };

  const { boxProps, Element } = useBox<React.ElementType>({
    ...otherProps,
    as: as || GRID_DEFAULT_TAG,
  });

  return (
    <Element
      {...boxProps}
      className={cx(
        css({
          display: isInline ? 'inline-grid' : 'grid',
          gridTemplateColumns: handleGridTemplate(columns),
          gridTemplateRows: handleGridTemplate(rows),
          flow,
          justifyContent,
          alignContent,
          columnGap: convertSpacingToToken(columnGap) ?? 0,
          rowGap: convertSpacingToToken(rowGap) ?? 0,
        }),
        boxProps.className,
      )}
      ref={ref}
    >
      {children}
    </Element>
  );
}

export const Grid: PolymorphicComponent<
  ExpandProps<GridInternalProps>,
  typeof GRID_DEFAULT_TAG
> = React.forwardRef(_Grid);
