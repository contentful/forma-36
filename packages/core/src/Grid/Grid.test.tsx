import React from 'react';
import { render, screen } from '@testing-library/react';

import { axe } from '../../../../scripts/test/axeHelper';
import { Grid } from './Grid';

it('renders the component', () => {
  const { container } = render(<Grid>Grid</Grid>);

  expect(container.firstChild).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const { container } = render(<Grid className="my-extra-class">Grid</Grid>);

  expect(container.firstChild).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const { container } = render(<Grid>Grid</Grid>);
  const results = await axe(container);

  expect(results).toHaveNoViolations();
});

describe('should have correct styles', () => {
  it('rowGap should be 0 by default', () => {
    render(<Grid>Grid</Grid>);
    const element = screen.getByText('Grid');

    expect(element).toMatchSnapshot();
    expect(screen.getByText('Grid')).toHaveStyle({ rowGap: 0 });
  });

  it('columnGap should be 1rem by default', () => {
    render(<Grid>Grid</Grid>);
    const element = screen.getByText('Grid');

    expect(screen.getByText('Grid')).toHaveStyle({ columnGap: '1rem' });
    expect(element).toMatchSnapshot();
  });

  it('should has correct numbers of columns', () => {
    render(<Grid columns={12}>Grid</Grid>);
    const element = screen.getByText('Grid');

    expect(element).toMatchSnapshot();
  });

  it('should has correct numbers of rows', () => {
    render(<Grid rows={12}>Grid</Grid>);
    const element = screen.getByText('Grid');

    expect(element).toMatchSnapshot();
  });
});
