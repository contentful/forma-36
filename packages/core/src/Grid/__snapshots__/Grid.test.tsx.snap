// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: auto;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="emotion-0"
>
  Grid
</div>
`;

exports[`renders the component with an additional class name 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: auto;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="my-extra-class emotion-0"
>
  Grid
</div>
`;

exports[`should have correct styles columnGap should be 1rem by default 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: auto;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="emotion-0"
>
  Grid
</div>
`;

exports[`should have correct styles rowGap should be 0 by default 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: auto;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="emotion-0"
>
  Grid
</div>
`;

exports[`should have correct styles should has correct numbers of columns 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: repeat(12,minmax(0,1fr));
  grid-template-rows: auto;
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="emotion-0"
>
  Grid
</div>
`;

exports[`should have correct styles should has correct numbers of rows 1`] = `
.emotion-0 {
  display: grid;
  grid-template-columns: auto;
  grid-template-rows: repeat(12,minmax(0,1fr));
  -webkit-column-gap: 1rem;
  column-gap: 1rem;
  row-gap: 0;
}

<div
  class="emotion-0"
>
  Grid
</div>
`;
